api/app/routers.py

```diff
@@ -1,15 +1,80 @@
+import logging
 import random
+from enum import Enum
 
 from django.conf import settings
+from django.core.cache import cache
+from django.db import connections
+
+logger = logging.getLogger(__name__)
+CONNECTION_CHECK_CACHE_TTL = 2
+
+
+class ReplicaReadStrategy(Enum):
+    DISTRIBUTED = "DISTRIBUTED"
+    SEQUENTIAL = "SEQUENTIAL"
+
+
+class ImproperlyConfiguredError(RuntimeError):
+    pass
+
+
+def connection_check(database: str) -> bool:
+    try:
+        conn = connections.create_connection(database)
+        conn.connect()
+        usable = conn.is_usable()
+        if not usable:
+            logger.warning(
+                f"Unable to access database {database} during connection check"
+            )
+    except Exception:
+        usable = False
+        logger.error(
+            "Encountered exception during connection",
+            exc_info=True,
+        )
+
+    if usable:
+        cache.set(
+            f"db_connection_active.{database}", "online", CONNECTION_CHECK_CACHE_TTL
+        )
+    else:
+        cache.set(
+            f"db_connection_active.{database}", "offline", CONNECTION_CHECK_CACHE_TTL
+        )
+
+    return usable
 
 
 class PrimaryReplicaRouter:
     def db_for_read(self, model, **hints):
         if settings.NUM_DB_REPLICAS == 0:
             return "default"
-        return random.choice(
-            [f"replica_{i}" for i in range(1, settings.NUM_DB_REPLICAS + 1)]
+
+        replicas = [f"replica_{i}" for i in range(1, settings.NUM_DB_REPLICAS + 1)]
+        replica = self._get_replica(replicas)
+        if replica:
+            # This return is the most likely as replicas should be
+            # online and properly functioning.
+            return replica
+
+        # Since no replicas are available, fall back to the cross
+        # region replicas which have worse availability.
+        cross_region_replicas = [
+            f"cross_region_replica_{i}"
+            for i in range(1, settings.NUM_CROSS_REGION_DB_REPLICAS + 1)
+        ]
```

Comment:
Nit: the `cross_region_replica_` and `replica_` prefixes could be constants used both by the router and the logic in settings module.

Reply:
I see what you're saying, but I don't think it's worth the legibility loss / added complexity.